pkg/codec/opus/opus.go

```diff
@@ -58,47 +61,98 @@ func newEncoder(r audio.Reader, p prop.Media, params Params) (codec.ReadCloser,
 		return nil, err
 	}
 
-	inBuffSize := int(targetLatency * float64(p.SampleRate) / 1000)
-	inBuff := wave.NewFloat32Interleaved(
-		wave.ChunkInfo{Channels: channels, Len: inBuffSize},
-	)
-	inBuff.Data = inBuff.Data[:0]
-	e := encoder{engine, inBuff, r}
+	e := encoder{
+		engine:         engine,
+		latencySamples: int(targetLatency * float64(p.SampleRate) / 1000),
+		channels:       channels,
+		reader:         r,
+		mixer:          params.ChannelMixer,
+	}
+	if e.mixer == nil {
+		e.mixer = &MonoMixer{}
+	}
 	return &e, nil
 }
 
-func (e *encoder) Read(p []byte) (n int, err error) {
+func (e *encoder) Read(p []byte) (int, error) {
+	nLatency := e.latencySamples * e.channels
+
 	// While the buffer is not full, keep reading so that we meet the latency requirement
-	nLatency := e.inBuff.ChunkInfo().Len * e.inBuff.ChunkInfo().Channels
-	for len(e.inBuff.Data) < nLatency {
+L_READ:
+	for {
+		switch inBuff := e.inBuff.(type) {
+		case *wave.Float32Interleaved:
+			if len(inBuff.Data) >= nLatency {
+				break L_READ
+			}
+		case *wave.Int16Interleaved:
+			if len(inBuff.Data) >= nLatency {
+				break L_READ
+			}
+		}
+
 		buff, err := e.reader.Read()
 		if err != nil {
 			return 0, err
 		}
-		// TODO: convert audio format
-		b, ok := buff.(*wave.Float32Interleaved)
-		if !ok {
-			panic("unsupported audio format")
-		}
-		switch {
-		case b.Size.Channels == 1 && e.inBuff.ChunkInfo().Channels != 1:
-			for _, d := range b.Data {
-				for ch := 0; ch < e.inBuff.ChunkInfo().Channels; ch++ {
-					e.inBuff.Data = append(e.inBuff.Data, d)
-				}
+		switch b := buff.(type) {
+		case *wave.Float32Interleaved:
```

Comment:
I think we can move this logic to `io/audio` and make it similar to how we do it for video scaling. So, before we read from the reader, we transform it so that it will always return the right latency.

Reply:
Do you have a good English word to express this functionality that splits signal at uniform interval?

Reply:
hmm, not really sure either... How about `Buffer`? Since we would put in under `io/audio` and the signature should be clear that this is a transformation, I think it's clear for the caller. e.g.

```go
import "github.com/pion/mediadevices/pkg/io/audio"

// This will create a reader with a buffer of 20 millisecond for each read
r, err := audio.Buffer(time.Millisecond * 20)
```